Adding/rearranging animations pages

Sfshaza opened the pull request:
This PR:
- Adds a new landing page
- Moves the old landing page to the new /animations subdirectory and calls it a tech overview
- Add's Jim Beveridge's tweening doc (with his permission)

Staged: https://flutter-io-deploy-two.firebaseapp.com/animations/

Please review @InMatrix @Hixie

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+In Flutter, an Animation object knows nothing about what is onscreen.
+An Animation is an abstract class that understands its current value
+and its state (completed or dismissed). One of the more commonly used
+animation types is Animation<double>.
```
here and elsewhere (e.g. the header), you need to put backticks around class names otherwise the `<` characters end up interpreted as HTML.

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+
+A few facts about Flutter animations:
+
+* An Animation object in Flutter is a class that sequentially generates
```
this is specifically talking about `Animation<double>`

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+an animation with the `.forward()` method. The generation of numbers is
+tied to the screen refresh, so typically 60 numbers are generated per
+second. After each number is generated, each Animation object calls the
+attached Listener objects.
```
i would just say "attached listeners", since they're usually closures or methods, not objects (except in the pedantic sense that closures, methods, and functions are objects).

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+
+When creating an AnimationController, you pass it a `vsync` argument.
+This option keeps the current widget tree in display memory until
+Flutter's rendering engine completes its refresh cycle. To use your
```
The sentence "This option keeps the current widget tree in display memory until Flutter's rendering engine completes its refresh cycle." is not at all correct. :-)

What vsync does is it ties the AnimationController to a particular place in the widget tree, so that when that part of the widget tree isn't on the screen, the AnimationController stops calling its callback (though without pausing the clock, so time still passes). Specifically, the vsync argument specifies a TickerProvider.

InMatrix replied:
What is that "particular place" in the widget tree? Is it the State object from which the TickerProvider is obtained?

Hixie replied:
It ties it to the TickerProvider; typically that's the State that mixes in the TickerProviderStateMixin. TickerProviderStateMixin further defers to the subtree that's rooted at the nearest TickerMode widget.

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+but that’s not a requirement.
+
+A Tween inherits from Animatable<T>, not from Animation<T>.
+An Animatable does not have to output double.
```
Nor does an `Animation<T>` (it produces a T). I'd probably move this sentence elsewhere lest people think that it is the explanation of the difference between Animation and Animatable.

Hixie replied:
if you do want to explain the difference here, it's what you say in the paragraph below -- Animations have state and a current value and can be listened to, while Animatables describe how to map an `Animation<double>`'s value to another value.

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+A Listener is called whenever the value of the animation changes.
+The most common behavior of a Listener is to call `setState()`
+to cause a rebuild. A StatusListener is called when an animation begins,
+ends, moves forward, or moves reverse, as defined by AnimationStatus.
```
might be simpler to describe this as "starts, stops, or changes direction"

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+
+The examples so far show how to generate a sequence of numbers over time.
+Nothing has been rendered to the screen.
+To render with an Animation<> object, store the Animation object as a
```
`<>`?

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+        margin: new EdgeInsets.symmetric(vertical: 10.0),
+        height: [[highlight]]animation.value,[[/highlight]]
+        width: [[highlight]]animation.value,[[/highlight]]
+        child: new FlutterLogo,
```
missing parens on the constructor here

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+The difference is that, in the first case,
+the animation variable hasn't been set when the closure is created.
+Therefore, if you need to use `animation.value` in `setState()`,
+you must use the second form.
```
it's been set by the time the closure runs, which should be sufficient.
maybe you mean that it hasn't been declared (and is thus out of scope)? If so, the example should use "var", "final", or a type on the assignment line to make it clear that that's where the variable is defined. (In this example it's not defined there, it's a member variable).

InMatrix commented on animations/index.md:
```diff
@@ -0,0 +1,115 @@
+behavior. When you toss a ball, for example, where and when
+it lands depends on how fast it was tossed, how heavy it is, and how
+far off the ground. Similarly, dropping a ball attached to a spring falls
+(and bounces) differently than dropping a ball attached to a string.
```
Should we point advanced users to [AnimationController.animateWith](https://docs.flutter.io/flutter/animation/AnimationController/animateWith.html) and [SpringSimulation](https://docs.flutter.io/flutter/physics/SpringSimulation-class.html)? We can acknowledge that we're working on docs for physics-based animations.

InMatrix commented on animations/index.md:
```diff
@@ -0,0 +1,115 @@
+and Flutter's animation architecture.
+
+* The [animation
+library](https://docs.flutter.io/flutter/animation/animation-library.html)
```
The landing page of animation library is not very useful as it is today. For one thing, many classes useful for building animations are outside of this library, e.g., [SizeTransition](https://docs.flutter.io/flutter/widgets/SizeTransition-class.html), etc. I'll file a bug to add more info to the animation library landing page. For the time being, I'd recommend link to https://flutter.io/widgets/animation/

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+      }
+    });
+    controller.forward();
+  }
```
you can also get this same effect with an appropriately configured sawtooth tween and "repeat" on the animation controller.

InMatrix commented on animations/index.md:
```diff
@@ -0,0 +1,115 @@
+you may also want to reference the
+[Navigator](https://docs.flutter.io/flutter/widgets/Navigator-class.html) and
+[PageRoute](https://docs.flutter.io/flutter/widgets/PageRoute-class.html)
+widgets.
```
Mention these classes where we talked about hero animation?

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,727 @@
+objects. Widgets can either incorporate these animations in their build
+functions directly by reading their current value and listening to their
+state changes or they can use the animations as the basis of more elaborate
+animations that they pass along to other widgets.
```
When we say "animation" here, it'd be helpful to distinguish between the general concept of animation and Dart objects with the Animation type.

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+which you’ll find in a helper class named AnimatedWidget.
+The simplified code is shown below. LogoApp now derives from
+AnimatedWidget instead of from StatefulWidget, and your
+AnimatedWidget does not need to maintain a State object to
```
But the parent of the AnimatedWidget needs to maintain a State object to hold the animation. I'm not sure if we should explain that.

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+it works exactly the same as before.
+
+You can find the source for this example,
+[animate2,](https://raw.githubusercontent.com/InMatrix/animation_tutorial/master/lib/animate2.dart)
```
Should I move the code to a place in the Flutter repo?

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+      [[highlight]]..addStatusListener((state) => print("$state"));[[/highlight]]
+    controller.forward();
+  }
+  //...
```
Can we make `//...` stand out a bit more? It's easy to copy and paste it and believe it's a self-contained code snippet.

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+is stateless and holds the set of final variables necessary to
+define the transition animation. The build() function creates and
+returns the AnimatedBuilder, which takes the (Anonymous builder)
+method and the LogoWidget object as parameters. The work of
```
LogoWidget object is indirectly passed into the (Anonymous builder) method through the child property of GrowTransition.

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+}
+{% endprettify %}
+
+The middle three blocks in the diagram are all created in the
```
I'm a little confused by the "ContainerTransition" in the diagram. It's not in the code. Is it the same thing as GrowTransition?

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+
+There are many more classes to explore, including specialized Tween
+classes, animations specific to Material Design, ReverseAnimation,
+physics simulations and `fling()` methods. This doc should have
```
Maybe we can link to relevant API docs here. Some developers might be able to figure things out, and if not, hopefully they will ask about them and file issues.

InMatrix commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+such as cascading events that delay the start of Animation B until three
+seconds after Animation A starts. Please contact abarth@ if you have
+use cases.
+{% endcomment %}
```
Remove this comment?

Hixie commented on animations/tween-animations.md:
```diff
@@ -0,0 +1,726 @@
+<!-- skip -->
+{% prettify dart %}
+final CurvedAnimation curve =
+    new CurvedAnimation(parent: controller, curve: Curves.easeIn);
```
Yes, it's not uncommon to chain multiple Animations.

It's the parent because that's the one that will drive this one.

Sfshaza commented:
I've incorporated @InMatrix's feedback on the landing page. I still have to incorporate @Hixie's feedback and update the converted tutorial. This is still in process.

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+The following resources are a good place to start learning the Flutter
+animation framework. Each of these docs shows, step by step, how
+to write animation code.
+Note that more docs are in the works on how to implement common design patterns,
```
nit: i would omit the words "Note that" from this sentence.

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+If you have a specific request, please
+[file an issue](https://github.com/flutter/flutter/issues).
+
+* [Tutorial: Animations in Flutter](/tutorials/animation/) <img src="/images/ic_new_releases_black_24px.svg" alt="this doc is new!"> NEW<br>
```
The alt text on this image is redundant with the text that follows it. We should have one or the other, not both.

Sfshaza replied:
Travis fails unless there's ALT text, or I would have left it off. The visible text is important to clarify the meaning of the icon.

Hixie replied:
We should fix Travis. It's invalid HTML to have redundant alt="" text. See https://html.spec.whatwg.org/multipage/images.html#alt

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+
+Animations fall into one of two categories: tween- or physics-based.
+The following sections explain what these terms mean, and points you to
+resources where you can learn more. Note that, in some cases,
```
nit: omit "Note that"

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+The following sections explain what these terms mean, and points you to
+resources where you can learn more. Note that, in some cases,
+the best documentation we currently have is example code in the
+Flutter gallery. More docs and examples will be forthcoming.
```
nit: here and elsewhere, maybe replace "docs" with "documents"?

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+far off the ground. Similarly, dropping a ball attached to a spring falls
+(and bounces) differently than dropping a ball attached to a string.
+
+Note that more docs and examples on physics-based animations will be
```
nit: i would remove "note that".

We may want to remove the promises for more documentation in general, actually. If we add more docs, great. If we never do, then these promises aren't helpful.

Hixie commented on animations/index.md:
```diff
@@ -0,0 +1,158 @@
+In this pattern, the user selects an element&mdash;often an
+image&mdash;from the page, and the UI animates the selected element
+to a new page with more detail. In Flutter, you can easily implement
+shared element transitions between routes (pages) using the Hero widget.
```
would it make sense, here and elsewhere, to link the class names to the API docs for those classes? Or is it better to only have the "also see" links below?

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+
+When creating an AnimationController, you pass it a `vsync` argument.
+This option keeps the current widget tree in display memory until
+Flutter's rendering engine completes its refresh cycle. To use your
```
The sentence "This option keeps the current widget tree in display memory until Flutter's rendering engine completes its refresh cycle." is wrong.

What 'vsync' does is tie the ticking of the animation controller to the visibility of the widget, so that when the animating widget goes off-screen, the ticking stops, and when the widget is restored, it starts again (without stopping the clock, so it's as if it had been ticking the whole time, but without using CPU).

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+velocity, force, and position (via the Force object).
+The position can be anything and so can be outside of the 0.0 to 1.0 range.
+
+A CurvedAnimation can also exceed the 0.0 to 1.0 range,
```
at this point you haven't introduced CurvedAnimation yet.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+
+CurvedAnimation and AnimationController are both of type
+Animation&lt;double&gt;, so you can pass them interchangeably.
+Note that the CurvedAnimation wraps the object it’s modifying.
```
nit: here and elsewhere, remove the words "note that" whenever they occur. They don't add anything.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+but that’s not a requirement.
+
+A Tween inherits from Animatable&lt;T&gt;, not from Animation&ltT&gt;.
+An Animatable does not have to output double.
```
> An Animatable does not have to output double.

Nor does an Animation. (It sounds like you're saying this as a contrast to Animation.)

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+{% endprettify %}
+
+A Tween object does not store any state. Instead, it exposes two methods
+`evaluate(Animation<double> animation)`, which applies the mapping function
```
you say two methods but only list one

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+#### Tween.animate
+
+To use a Tween object, you configure an Animation to use the Tween
+(Animatable) with the `animate()` function on the Animatable object.
```
this sentence is convoluted.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+The following shows the same code modified to animate the
+logo to grow from nothing to full size. Note that when
+defining an AnimationController, you must pass in a `vsync` object.
+The presence of `vsync` prevents offscreen animations from consuming
```
this definition of vsync is accurate. use it in the other place too if you need it in both places. :-)

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+      [[highlight]]});[[/highlight]]
+    [[highlight]]controller.forward();[[/highlight]]
+  [[highlight]]}[[/highlight]]
+
```
this demo leaks the animation controller. You need a dispose method where you call animation.dispose().

Sfshaza replied:
Thanks! The source code is under Tao's repo. I've asked him to fix.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+{% endprettify %}
+
+The difference is that, in the first case,
+the animation variable hasn't been set when the closure is created.
```
that doesn't matter. What matters is whether the variable is in scope, and it is, so both techniques would work fine.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+The difference is that, in the first case,
+the animation variable hasn't been set when the closure is created.
+Therefore, if you need to use `animation.value` in `setState()`,
+you must use the second form.
```
that is, this last sentence here is wrong.

Sfshaza replied:
I deleted those sentences. thx

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+[animate5](https://raw.githubusercontent.com/InMatrix/animation_tutorial/master/lib/animate5.dart)
+example looks like this:
+
+<img src="images/AnimatedBuilder-WidgetTree.png" alt="a widget tree with Container pointing to ContainerTransition, pointing to AnimatedBuilder, pointing to (AnonymousBuilder), pointing to LogoWidget">
```
Capitalize the first word of the alt text, finish it with a period.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+  Widget build(BuildContext context) {
+    return new AnimatedBuilder(animation: animation,
+        builder: (BuildContext context, Widget child) {
+          return new Center(
```
the center could be placed outside the AnimatedBuilder to be more efficient.

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+  Animation animation;
+  AnimationController controller;
+
+  initState() {
```
`@override`

Hixie commented on tutorials/animation/index.md:
```diff
@@ -0,0 +1,874 @@
+  @override
+  dispose() {
+    super.dispose();
+    controller.dispose();
```
here and elsewhere: convention is to call `super.dispose()` last, since in principle once you've called it your object is dead and you shouldn't do anything anymore.

Sfshaza commented:
Thanks, Ian, for your thorough review! I believe I've fixed everything, except where noted. I removed all @overrides from the examples, since the analyzer is happy without. I've sent Tao email asking him to update the code on GitHub, with a list of things to update. (I have tested all the updated code, but it's on my hard drive, not GitHub.)

Sfshaza commented:
The dupe _code is because I want to move the layout examples under a layout directory, but Travis will fail if I don't do this as a 2-step process, because it checks links. Once the code is in place, I can submit another PR to change the links and delete the old cold.

Sfshaza commented:
With Devon's fix, Travis is happy again. @Hixie, can this PR be pushed?

Hixie commented:
LGTM.
